src/RevPrefetcher.cc

```diff
@@ -64,6 +74,17 @@ bool RevPrefetcher::IsAvail(uint64_t Addr){
   return false;
 }
 
+void RevPrefetcher::MarkInstructionLoadComplete(MemReq& req){
+  auto it = OutstandingFetchQ.begin();
+  while(!OutstandingFetchQ.empty() && (it != OutstandingFetchQ.end())){
```

Comment:
Is testing `OutstandingFetchQ.empty()` necessary? I thought that if it were empty, `it == OutstandingFetchQ.end()`.

Reply:
It appeared to be running off the end without that check

Reply:
does `LSQHash()` return `LSQueue.end()`? That might be why the `empty()` is required.

Reply:
`LSQHash()`, whether as the namespace function accepting 3 parameters, or the `MemReq` member function depending on 3 `MemReq` properties, returns a `uint64_t` hash, which if used as a key in `find()` for a `std::unordered_multimap`, returns an iterator which should equal `end()` if there was no match.

https://github.com/tactcomplabs/rev/blob/1468d6d672bd69abe2f33273c13fb9455ee76a7c/common/include/RevCommon.h#L88

In this case `begin()` is used, which should equal `end()` if the map is empty. You can erase entries, which invalidates the erased iterator and its pointed-to-data, but the next iterator should be okay, even if it equals `end()`. It might be necessary to compute the next iterator before `erase()` is called on the current iterator.

Reply:
```c++
OutstandingFetchQ.erase(it++);
```
Instead of:
```c++
OutstandingFetchQ.erase(it);
it++;
```
... because `it++`, which is executed before `erase()` is called, will set `it` to the next iterator before `erase()` is called, and then even if `erase()` invalidates the old value of `it` and what it pointed to, `it` will already be incremented to the next value before `erase()` is called, and will be set to `end()` if the map is finished being traversed.

Reply:
fixed

Reply:
Think of deleting a linked list entry, whose next  entry's iterator is computed from a `next` pointer stored in the current entry. By calling `erase(it++)` you are computing the `next` entry and setting `it` to point to it, before calling `erase(it)` on the old value of `it` and destroying the `next` pointer store in `*it`.